corehq/apps/reports/standard/users/reports.py

```diff
@@ -116,21 +149,56 @@ def rows(self):
             self.pagination.start:self.pagination.start + self.pagination.count
         ]
         for record in records:
-            yield _user_history_row(record, self.timezone)
+            yield _user_history_row(record, self.domain, self.timezone)
 
 
-def _user_history_row(record, timezone):
+def _user_history_row(record, domain, timezone):
     return [
         cached_user_id_to_username(record.user_id),
         cached_user_id_to_username(record.changed_by),
         _get_action_display(record.action),
         record.details['changed_via'],
         record.message,
-        json.dumps(record.details['changes']),
+        _details_cell(record.details['changes'], domain),
         ServerTime(record.changed_at).user_time(timezone).ui_string(USER_DATETIME_FORMAT),
     ]
 
 
+def _details_cell(changes, domain):
+    def _html_list(changes, unstyled=True):
+        items = []
+        for key, value in changes.items():
+            if isinstance(value, dict):
+                value = _html_list(value, unstyled=unstyled)
+            elif isinstance(value, list):
+                value = format_html(", ".join(value))
+            else:
+                value = format_html(str(value))
+            items.append("<li>{}: {}</li>".format(key, value))
+
+        class_attr = "class='list-unstyled'" if unstyled else ""
+        return f"<ul {class_attr}>{''.join(items)}</ul>"
+
+    properties = get_primary_properties(domain)
+    properties.pop("user_data", None)
+    primary_changes = {
+        properties.get(key, key): value for key, value in changes.items()
+        if key in properties
+    }
+    more_count = len(changes) - len(primary_changes)
+    if more_count == 0:
+        return mark_safe(_html_list(primary_changes))
+
+    see_more = _("See {count} more").format(count=more_count)
+    return mark_safe(f"""
+        <div>
+          <div>{_html_list(primary_changes) if primary_changes else ""}</div>
+          <a class="see-all-link">{see_more}</a>
+          <div class="hide">{_html_list(changes, unstyled=False)}</div>
+        </div>
+    """)
```

Comment:
Is it possible to put this in a html template and render that with context instead?
This method has grown to be quite complex and then is handling html render as well.

Reply:
4cbabbf6e58

Reply:
thank you :)